pkg/registry/apis/provisioning/repository_controller.go

```diff
@@ -0,0 +1,226 @@
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	provisioning "github.com/grafana/grafana/pkg/apis/provisioning/v0alpha1"
+	client "github.com/grafana/grafana/pkg/generated/clientset/versioned/typed/provisioning/v0alpha1"
+	informer "github.com/grafana/grafana/pkg/generated/informers/externalversions/provisioning/v0alpha1"
+	listers "github.com/grafana/grafana/pkg/generated/listers/provisioning/v0alpha1"
+	"github.com/grafana/grafana/pkg/registry/apis/provisioning/auth"
+)
+
+// RepositoryController controls how and when CRD is established.
+type RepositoryController struct {
+	client     client.ProvisioningV0alpha1Interface
+	repoLister listers.RepositoryLister
+	repoSynced cache.InformerSynced
+
+	// Converts config to instance
+	repoGetter RepoGetter
+	logger     *slog.Logger
+	identities auth.BackgroundIdentityService
+
+	// To allow injection for testing.
+	syncFn            func(key string) error
+	enqueueRepository func(obj any)
+	keyFunc           func(obj any) (string, error)
+
+	queue workqueue.TypedRateLimitingInterface[string]
+}
+
+// NewRepositoryController creates new RepositoryController.
+func NewRepositoryController(
+	provisioningClient client.ProvisioningV0alpha1Interface,
+	repoInformer informer.RepositoryInformer,
+) (*RepositoryController, error) {
+	rc := &RepositoryController{
+		client:     provisioningClient,
+		repoLister: repoInformer.Lister(),
+		repoSynced: repoInformer.Informer().HasSynced,
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.DefaultTypedControllerRateLimiter[string](),
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: "provisioningRepositoryController"},
+		),
+	}
+
+	_, err := repoInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    rc.addRepository,
+		UpdateFunc: rc.updateRepository,
+		DeleteFunc: rc.deleteRepository,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	rc.syncFn = rc.sync
+	rc.enqueueRepository = rc.enqueue
+	rc.keyFunc = repoKeyFunc
+
+	return rc, nil
+}
+
+func repoKeyFunc(obj any) (string, error) {
+	repo, ok := obj.(*provisioning.Repository)
+	if !ok {
+		return "", fmt.Errorf("expected a Repository but got %T", obj)
+	}
+	return cache.DeletionHandlingMetaNamespaceKeyFunc(repo)
+}
+
+// Run starts the RepositoryController.
+func (rc *RepositoryController) Run(ctx context.Context, workerCount int) {
+	defer utilruntime.HandleCrash()
+	defer rc.queue.ShutDown()
+	logger := klog.FromContext(ctx)
```

Comment:
why using `klog` and not `slog.Logger`?

Reply:
it can be updated. a lot of the boilerplate in this file was based in controllers from `kubernetes/kubernetes`

Reply:
I updated it